src/coreclr/vm/riscv64/stubs.cpp

```diff
@@ -1009,55 +1009,67 @@ BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDes
-            }
-        }
-        else
-        {
-            Emit32((DWORD)(0x00000013 | (constant & 0xFFF) << 20 | (target << 7))); // addi target, x0, constant
-        }
-    }
-    else
+    // Adaptation of emitLoadImmediate
+
+    if (isValidSimm12(imm))
     {
-        UINT32 upper = constant >> 32;
-        if (((upper + 0x800) >> 12) != 0)
-        {
-            Emit32((DWORD)(0x00000037 | (((upper + 0x800) >> 12) << 12) | (target << 7))); // lui target, (upper + 0x800) >> 12
-            if ((upper & 0xFFF) != 0)
-            {
-                Emit32((DWORD)(0x00000013 | (upper & 0xFFF) << 20 | (target << 7) | (target << 15))); // addi target, target, upper 
-            }
-        }
-        else
-        {
-            Emit32((DWORD)(0x00000013 | (upper & 0xFFF) << 20 | (target << 7))); // addi target, x0, upper 
-        }
-        UINT32 lower = (constant << 32) >> 32;
-        UINT32 shift = 0;
-        for (int i = 32; i >= 0; i -= 11)
-        {
-            shift += i > 11 ? 11 : i;
-            UINT32 current = lower >> (i < 11 ? 0 : i - 11);
-            if (current != 0)
-            {
-                Emit32((DWORD)(0x00001013 | (shift << 20) | (target << 7) | (target << 15))); // slli target, target, shift
-                Emit32((DWORD)(0x00000013 | (current & 0x7FF) << 20 | (target << 7) | (target << 15))); // addi target, target, current
-                shift = 0;
-            }
-        }
-        if (shift)
-        {
-            Emit32((DWORD)(0x00001013 | (shift << 20) | (target << 7) | (target << 15))); // slli target, target, shift
+        EmitAddImm(reg, 0 /* zero register */, imm & 0xFFF);
+        return;
+    }
+
+    // TODO-RISCV64: maybe optimized via emitDataConst(), check #86790
+
+    UINT32 msb;
+    UINT32 high31;
+
+    BitScanReverse64(&msb, imm);
+
+    if (msb > 30) {
+        high31 = (imm >> (msb - 30)) & 0x7FffFFff;
+    } else {
+        high31 = imm & 0x7FffFFff;
+    }
+
+    // Since ADDIW use sign extension for immediate
+    // we have to adjust higher 19 bit loaded by LUI
+    // for case when low part is bigger than 0x800.
+    UINT32 high19 = (high31 + 0x800) >> 12;
+
+    EmitLuImm(reg, high19);
+    if (high31 & 0x800) {
+        // EmitAddImm does not allow negative immediate values, so use EmitSubImm.
+        EmitSubImm(reg, reg, ~high31 + 1 & 0xFFF);
+    } else {
+        EmitAddImm(reg, reg, high31 & 0x7FF);
+    }
+
+    // And load remaining part by batches of 11 bits size.
+    INT32 remainingShift = msb - 30;
+
+    // shiftAccumulator usage is an optimization allows to exclude `slli addi` iteration
+    // if immediate bits `low11` for this iteration are zero.
+    UINT32 shiftAccumulator = 0;
+
+    while (remainingShift > 0) {
+        UINT32 shift = remainingShift >= 11 ? 11 : remainingShift % 11;
+        UINT32  mask  = 0x7ff >> (11 - shift);
+        remainingShift -= shift;
+        ssize_t low11 = (imm >> remainingShift) & mask;
```

Comment:
```suggestion
        UINT32 low11 = (imm >> remainingShift) & mask;
```
Everything else in this method is UINT32. Is there a reason for `ssize_t` here?

Reply:
Thank you!